Research/QuantBook.cs

```diff
@@ -144,62 +172,144 @@ public QuantBook() : base()
+        /// <param name="selector">Selects a value from the Fundamental data to filter the request output</param>
+        /// <param name="start">The start date of selected data</param>
+        /// <param name="end">The end date of selected data</param>
+        /// <returns>Enumerable collection of DataDictionaries, one dictionary for each day there is data</returns>
+        public IEnumerable<DataDictionary<dynamic>> GetFundamental(IEnumerable<Symbol> symbols, string selector, DateTime? start = null, DateTime? end = null)
+        {
+            var data = GetAllFundamental(symbols, start, end);
 
-                foreach (var ticker in tickers)
+            // Convert the data to proper form; a list of DataDictionaries to represent a day
+            var result = new List<DataDictionary<dynamic>>();
+            foreach (var kvp in data)
+            {
+                foreach (var value in kvp.Value)
                 {
-                    var symbol = QuantConnect.Symbol.Create(ticker.ToString(), SecurityType.Equity, Market.USA);
-                    var dir = new DirectoryInfo(Path.Combine(Globals.DataFolder, "equity", symbol.ID.Market, "fundamental", "fine", symbol.Value.ToLowerInvariant()));
-                    if (!dir.Exists) continue;
+                    // Take the whole value, may need to be replaced
+                    dynamic valueToAdd = value;
 
-                    var config = new SubscriptionDataConfig(typeof(FineFundamental), symbol, Resolution.Daily, TimeZones.NewYork, TimeZones.NewYork, false, false, false);
+                    // Check our results to see if there is a dictionary for this day
+                    var dayDictionary = result.Find(x => x.Time == value.Time) ?? new DataDictionary<dynamic>(value.Time);
 
-                    foreach (var fileName in dir.EnumerateFiles())
+                    // Put the value in at that date; filter if there is a selector
+                    if (!selector.IsNullOrEmpty())
                     {
-                        var date = DateTime.ParseExact(fileName.Name.Substring(0, 8), DateFormat.EightCharacter, CultureInfo.InvariantCulture);
-                        if (date < start || date > end) continue;
+                        valueToAdd = GetPropertyValue(value, selector);
+                    }
 
-                        var factory = new TextSubscriptionDataSourceReader(_dataCacheProvider, config, date, false);
-                        var source = new SubscriptionDataSource(fileName.FullName, SubscriptionTransportMedium.LocalFile);
-                        var value = factory.Read(source).Select(x => GetPropertyValue(x, selector)).First();
+                    dayDictionary.Add(kvp.Key, valueToAdd);
 
-                        list.Add(Tuple.Create(symbol, date, value));
+                    // Add the dayDictionary to the results if it isn't already entered
+                    if (!result.Contains(dayDictionary))
```

Comment:
Let's circle back I think there is some room for improvement in this loop, specifically the `result.Find/Contains`. Maybe we could send into `GetAllFundamental` a data handler so we can specify different behavior for C#/Py so it's easier to return.